SpatialGDK/Source/SpatialGDK/Public/LoadBalancing/LayeredLBStrategy.h

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#pragma once
+
+#include "LoadBalancing/AbstractLBStrategy.h"
+
+#include "CoreMinimal.h"
+#include "Math/Box2D.h"
+#include "Math/Vector2D.h"
+
+#include "LayeredLBStrategy.generated.h"
+
+class SpatialVirtualWorkerTranslator;
+class UAbstractLockingPolicy;
+
+DECLARE_LOG_CATEGORY_EXTERN(LogLayeredLBStrategy, Log, All)
+
+USTRUCT()
+struct FLBLayerInfo
+{
+	GENERATED_BODY()
+
+	UPROPERTY()
+	FName Name;
+
+	UPROPERTY(EditAnywhere, Category = "Load Balancing")
+	TSubclassOf<UAbstractLBStrategy> LoadBalanceStrategy;
+
+	UPROPERTY(EditAnywhere, Category = "Load Balancing")
+	TSubclassOf<UAbstractLockingPolicy> LockingPolicy;
+
+
+	FLBLayerInfo() : Name(NAME_None)
+	{
+	}
+};
+
+/**
+ * A load balancing strategy that ...
+ */
+UCLASS(Blueprintable)
+class SPATIALGDK_API ULayeredLBStrategy : public UAbstractLBStrategy
+{
+	GENERATED_BODY()
+
+public:
+	ULayeredLBStrategy();
+	~ULayeredLBStrategy();
+
+	/* UAbstractLBStrategy Interface */
+	virtual void Init() override;
+
+	virtual void SetLocalVirtualWorkerId(VirtualWorkerId InLocalVirtualWorkerId) override;
+
+	virtual bool ShouldHaveAuthority(const AActor& Actor) const override;
+	virtual VirtualWorkerId WhoShouldHaveAuthority(const AActor& Actor) const override;
+
+	virtual SpatialGDK::QueryConstraint GetWorkerInterestQueryConstraint() const override;
+
+	virtual bool RequiresHandoverData() const { return GetMinimumRequiredWorkers() > 1; }
+
+	virtual FVector GetWorkerEntityPosition() const override;
+
+	virtual uint32 GetMinimumRequiredWorkers() const override;
+	virtual void SetVirtualWorkerIds(const VirtualWorkerId& FirstVirtualWorkerId, const VirtualWorkerId& LastVirtualWorkerId)  override;
+	/* End UAbstractLBStrategy Interface */
+
+protected:
+
+private:
+	mutable TMap<TSoftClassPtr<AActor>, FName> ClassPathToLayer;
+
+	TMap<VirtualWorkerId, FName> VirtualWorkerIdToLayerName;
+
+	TMap<FName, UAbstractLBStrategy* > LayerNameToLBStrategy;
+
+	// Returns the first Layer that contains this, or a parent of this class,
+	// or the default actor group, if no mapping is found.
+	FName GetLayerForClass(TSubclassOf<AActor> Class) const;
+
+	// Returns true if ActorA and ActorB are contained in Layers that are
+	// on the same Server worker type.
+	bool IsSameWorkerType(const AActor* ActorA, const AActor* ActorB) const;
```

Comment:
Wonder if this could use const AActor refs as most of your other functions in this class do?

Reply:
We need to support the pointers being null.